tests/components/test_gpio_pin.py

```diff
@@ -0,0 +1,358 @@
+"""Tests for the GPIO Pin Classes."""
+from typing import List, Type
+
+import pytest
+
+from j5.backends import Backend
+from j5.boards import Board
+from j5.components import Component, NotSupportedByHardware
+from j5.components.gpio_pin import (
+    BadGPIOPinMode,
+    GPIOPin,
+    GPIOPinInterface,
+    GPIOPinMode,
+)
+
+
+class MockGPIOPinDriver(GPIOPinInterface):
+    """A testing driver for the GPIO pin component."""
+
+    def __init__(self):
+
+        self.pin_count: int = 10
+        self._mode: List[GPIOPinMode] = [
+            GPIOPinMode.DIGITAL_OUTPUT for _ in range(0, self.pin_count)
+        ]
+
+        self._written_digital_state: List[bool] = [
+            False for _ in range(0, self.pin_count)
+        ]
+
+        self._digital_state: List[bool] = [
+            False for _ in range(0, self.pin_count)
+        ]
+
+    def set_gpio_pin_mode(self, board: Board, identifier: int, pin_mode: GPIOPinMode):
+        """Set the hardware mode of a pin."""
+        self._mode[identifier] = pin_mode
+
+    def get_gpio_pin_mode(self, board: Board, identifier: int) -> GPIOPinMode:
+        """Get the hardware mode of a GPIO pin."""
+        return self._mode[identifier]
+
+    def write_gpio_pin_digital_state(self, board: Board, identifier: int, state: bool):
+        """Write to the digital state of a GPIO pin."""
+        self._written_digital_state[identifier] = state
+
+    def get_gpio_pin_digital_state(self, board: Board, identifier: int):
+        """Get the last written state of the GPIO pin."""
+        return self._written_digital_state[identifier]
+
+    def read_gpio_pin_digital_state(self, board: Board, identifier: int):
+        """Read the digital state of the GPIO pin."""
+        return self._digital_state[identifier]
+
+    def read_gpio_pin_analogue_value(self, board: Board, identifier: int) -> float:
+        """Read the scaled analogue value of the GPIO pin."""
+        return 0.6
+
+    def write_gpio_pin_dac_value(
+            self,
+            board: Board,
+            identifier: int,
+            scaled_value: float,
+    ) -> None:
+        """Write a scaled analogue value to the DAC on the GPIO pin."""
+        pass
```

Comment:
Are these (this method and `write_gpio_pin_pwm_value`) methods not being implemented for testing intentionally? Fine if they are, just want to make sure they weren't accidentally missed.

Reply:
This is as intended. I have ideas about how to make unit testing better, but cannot do that until #93 is finished.